java/statestore/src/main/java/sleeper/statestore/transactionlog/DynamoDBTransactionLogStore.java

```diff
@@ -119,6 +120,31 @@ public Stream<TransactionLogEntry> readTransactionsAfter(long lastTransactionNum
                 .map(this::readTransaction);
     }
 
+    @Override
+    public void deleteTransactionsAtOrBefore(long transactionNumber, Instant updateTime) {
+        streamPagedItems(dynamo, new QueryRequest()
+                .withTableName(logTableName)
+                .withConsistentRead(true)
+                .withKeyConditionExpression("#TableId = :table_id AND #Number <= :number")
+                .withFilterExpression("#UpdateTime <= :updateTime")
+                .withExpressionAttributeNames(Map.of("#TableId", TABLE_ID, "#Number", TRANSACTION_NUMBER, "#UpdateTime", UPDATE_TIME))
+                .withExpressionAttributeValues(new DynamoDBRecordBuilder()
+                        .string(":table_id", sleeperTableId)
+                        .number(":number", transactionNumber)
+                        .number(":updateTime", updateTime.toEpochMilli())
+                        .build())
+                .withReturnConsumedCapacity(ReturnConsumedCapacity.TOTAL))
+                .forEach(item -> {
+                    long itemTransactionNumber = getLongAttribute(item, TRANSACTION_NUMBER, 0L);
+                    if (item.get(BODY_S3_KEY) != null) {
+                        LOGGER.info("Deleting transaction body in S3 for transaction number {} and table {}", itemTransactionNumber, sleeperTableId);
+                        s3.deleteObject(new DeleteObjectRequest(dataBucket, item.get(BODY_S3_KEY).getS()));
+                    }
+                    LOGGER.info("Deleting transaction from DynamoDB for transaction number {} and table {}", itemTransactionNumber, sleeperTableId);
+                    dynamo.deleteItem(logTableName, getKey(item));
+                });
```

Comment:
It might be worth using BatchWriteItem to delete the transactions. There can be hundreds or thousands of transactions to delete, and it may produce contention with other operations on DynamoDB.

It might even be worth limiting the rate at which we delete the transactions, though it might be worth assuming we don't need that to begin with.

Reply:
We probably don't need to worry about this yet, we don't know if it'll be a problem.